File: langs/javascript/adapter.js
#!/usr/bin/env node
"use strict";

const fs = require("fs");
const path = require("path");

function loadClass(file, className) {
  const mod = require(path.resolve(file));
  const cls = mod[className] || mod.default || mod;
  if (typeof cls !== "function") {
    throw new Error("missing class " + className);
  }
  return cls;
}

function main() {
  const file = process.argv[2];
  const className = process.argv[3];
  const casesPath = process.argv[4];
  const Cls = loadClass(file, className);
  const cases = JSON.parse(fs.readFileSync(casesPath, "utf8"));
  const failed = [];
  let passed = 0;
  for (const c of cases) {
    const obj = new Cls();
    let ok = true;
    for (let i = 0; i < c.calls.length; i++) {
      const call = c.calls[i];
      const name = call.m.includes("_")
        ? call.m
            .split("_")
            .map((p, i) => (i === 0 ? p : p[0].toUpperCase() + p.slice(1)))
            .join("")
        : call.m;
      const fn = obj[name];
      let actual;
      try {
        actual = fn.apply(obj, call.a);
      } catch (err) {
        failed.push({
          case: c.id,
          index: i,
          method: call.m,
          expected: call.e,
          actual: "exc:" + err.name,
        });
        ok = false;
        break;
      }
      const exp = call.e;
      const same = JSON.stringify(actual) === JSON.stringify(exp);
      if (!same) {
        failed.push({
          case: c.id,
          index: i,
          method: call.m,
          expected: exp,
          actual: actual,
        });
        ok = false;
        break;
      }
    }
    if (ok) passed += 1;
  }
  process.stdout.write(JSON.stringify({ passed, failed }) + "\n");
  process.exit(failed.length ? 1 : 0);
}

main();

File: langs/ruby/adapter.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'

def load_class(file, class_name)
  path = File.expand_path(file)
  dir = File.dirname(path)
  $LOAD_PATH.unshift(dir) unless $LOAD_PATH.include?(dir)
  require File.basename(path, '.rb')
  Object.const_get(class_name)
end

def main
  file = ARGV[0]
  class_name = ARGV[1]
  cases_path = ARGV[2]
  cls = load_class(file, class_name)
  cases = JSON.parse(File.read(cases_path))
  failed = []
  passed = 0
  cases.each do |c|
    obj = cls.new
    ok = true
    c['calls'].each_with_index do |call, i|
      method = call['m']
      args = call['a']
      expected = call['e']
      begin
        actual = obj.public_send(method, *args)
      rescue StandardError => err
        failed << {
          'case' => c['id'],
          'index' => i,
          'method' => method,
          'expected' => expected,
          'actual' => "exc:#{err.class.name}"
        }
        ok = false
        break
      end
      if JSON.generate(actual) != JSON.generate(expected)
        failed << {
          'case' => c['id'],
          'index' => i,
          'method' => method,
          'expected' => expected,
          'actual' => actual
        }
        ok = false
        break
      end
    end
    passed += 1 if ok
  end
  $stdout.write(JSON.generate({ 'passed' => passed, 'failed' => failed }) + "\n")
  exit(failed.empty? ? 0 : 1)
end

main
